omnimove_external_control/src/omnimove_external_control.cpp

```diff
@@ -0,0 +1,357 @@
+  for (size_t i = 0; i < read_buffer_.size(); ++i)
+  {
+    msg_data.get()[i] = read_buffer_[i];
+  }
+  if (!ExternalControlData::isMessageValid(msg_data.get(), read_buffer_.size()))
+  {
+    const char * EXTERNAL_CONTROL_DATA_HEADER = "KMRUTV03";
+
+    for (size_t i = 0; i < strlen(EXTERNAL_CONTROL_DATA_HEADER); ++i)
+    {
+      if (read_buffer_[i] != EXTERNAL_CONTROL_DATA_HEADER[i])
+      {
+        RCLCPP_ERROR(
+          rclcpp::get_logger("OmnimoveExternalControl"), "msg_data is %d, header is %d",
+          read_buffer_[i], EXTERNAL_CONTROL_DATA_HEADER[i]);
+      }
+    }
+
+    return ExternalControlData();
+  }
+  ExternalControlData parsedData(msg_data.get() + 8);
+  read_buffer_.erase(read_buffer_.begin(), read_buffer_.begin() + expected_data_size);
+  return parsedData;
+  // now we will connect the latest message
+
+  // ExternalControlData received_data(read_buffer_.);
+}
+hardware_interface::return_type OmnimoveExternalControl::read(
+  const rclcpp::Time &, const rclcpp::Duration &)
+{
+  std::array<char, 1024> buffer;
+  if (client_socket_.get() == NULL)
+  {
+    RCLCPP_WARN(rclcpp::get_logger("OmnimoveExternalControl"), "socket is null !!!");
+  }
+  //   RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Am waiting to read something if
+  //   available");
+  size_t bytes_received = client_socket_->read_some(boost::asio::buffer(buffer));
+  // RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "received %lu", bytes_received);
+
+  read_buffer_.insert(read_buffer_.end(), buffer.begin(), buffer.begin() + bytes_received);
+  // RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Inserted into buffer");
+
+  // parse all messages from the buffer.
+  // incomplete messages need to be kept for further reading.
+  ExternalControlData readData = parseLastMessageFromBuffer();
+  //  RCLCPP_INFO(rclcpp::get_logger("OmnimoveExternalControl"), "Finished parsing into buffer");
+
+  if (readData.isDataValid())
+  {
+    if (agv_type_ == "caterpillar")
+    {
+      velocity_state_[0] = (double)readData.speedX();
+      velocity_state_[1] = (double)readData.speedW();
+      position_state_[0] = (double)readData.posPillar1();
+      velocity_state_[2] = (double)readData.speedPillar1();
+      position_state_[1] = (double)readData.posPillar2();
+      velocity_state_[3] = (double)readData.speedPillar2();
+      position_state_[2] = (double)readData.posPillar3();
+      velocity_state_[4] = (double)readData.speedPillar3();
+      position_state_[3] = (double)readData.posPillar4();
+      velocity_state_[5] = (double)readData.speedPillar4();
+      position_state_[4] = (double)readData.posShield();
+      velocity_state_[6] = (double)readData.speedShield();
+    }
+    else
+    {
+      velocity_state_[0] = (double)readData.speedX();
+      velocity_state_[1] = (double)readData.speedY();
+      velocity_state_[2] = (double)readData.speedW();
+    }
```

Comment:
It is better practice to use static cast instead of C-style function-like casting
```suggestion
      velocity_state_[0] = static_cast<double>(readData.speedX());
      velocity_state_[1] = static_cast<double>(readData.speedW());
      position_state_[0] = static_cast<double>(readData.posPillar1());
      velocity_state_[2] = static_cast<double>(readData.speedPillar1());
      position_state_[1] = static_cast<double>(readData.posPillar2());
      velocity_state_[3] = static_cast<double>(readData.speedPillar2());
      position_state_[2] = static_cast<double>(readData.posPillar3());
      velocity_state_[4] = static_cast<double>(readData.speedPillar3());
      position_state_[3] = static_cast<double>(readData.posPillar4());
      velocity_state_[5] = static_cast<double>(readData.speedPillar4());
      position_state_[4] = static_cast<double>(readData.posShield());
      velocity_state_[6] = static_cast<double>(readData.speedShield());
    }
    else
    {
      velocity_state_[0] = static_cast<double>(readData.speedX());
      velocity_state_[1] = static_cast<double>(readData.speedY());
      velocity_state_[2] = static_cast<double>(readData.speedW());
    }
```